### PR TITLE
Support Shadow DOM and iframe element grabbing

### PR DESCRIPTION
**Pull request by aidenybai**
## Summary

- select the deepest accessible element across open Shadow DOM and same-origin iframe boundaries
- preserve composed ancestry, selectors, anchors, bounds, and drag selection across nested roots and scaled frames
- keep same-origin iframe scrolling responsive during selection while caching safe fallback behavior for inaccessible frames
- add realistic Pierre diff, Shadow DOM edge-case, and iframe preview fixtures with focused Playwright coverage

## Tests

- `nr build`
- `nr lint`
- `nr typecheck`
- `nr format`
- `nr test:unit` (106 passed)
- focused iframe and Shadow DOM Playwright suites (8 passed)

---

> [!NOTE]
> **High Risk**
> Touches core hit-testing, overlay bounds, cross-document event forwarding, and freeze behavior—high regression surface for selection, scrolling, and focus across nested roots.
> 
> **Overview**
> Extends **react-grab** so pointer hit-testing, selection, drag bounds, hierarchy, selectors, and element relinking work through **open Shadow DOM**, **nested slots**, and **same-origin iframes** (with scaled/transform coordinate mapping and inaccessible-frame fallback to the `iframe` element).
> 
> Same-origin child documents **reuse the parent `__REACT_GRAB__` instance**, register dynamically (including iframes inside shadow roots), and while grab is active **forward pointer/keyboard/viewport events** into the top window with converted coordinates; global **animation and pseudo-state freeze** now applies per registered document. `freeze` / `getElementsAtPosition` use the same deep stack as the overlay.
> 
> The Vite e2e app adds **iframe**, **Pierre diff**, and **shadow edge** fixtures plus large Playwright suites; existing e2e helpers are tightened (scoped todo drag targets, edit-panel typing/focus, copy retries, relaxed post-deactivate scroll assertion).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 60bb8a959976edd7ea764fe4aa6704a14a52b239. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/react-grab/src/utils/is-user-ignored-element.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { USER_IGNORE_ATTRIBUTE } from "../constants.js";
+import { getComposedParentElement } from "./get-composed-parent-element.js";
+
+export const isUserIgnoredElement = (element: Element): boolean => {
+  let currentElement: Element | null = element;
+  while (currentElement) {
+    if (currentElement.hasAttribute(USER_IGNORE_ATTRIBUTE)) return true;
+    currentElement = getComposedParentElement(currentElement);
+  }
+  return false;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/is-user-ignored-element.ts:10]: Checked this against the slotting model and a focused browser reproduction. A slotable is a direct child of its host, and the traversal is node → assigned slot → shadow root host, so the light host is still visited; traversal then continues through the host’s light-DOM ancestors. The existing ignored-slotted-target E2E test passes, and moving the ignore attribute onto the host also falls through to the underlying target. The same path reaches a scoped host/container. There are therefore no skipped intermediate light ancestors here.

File: packages/react-grab/src/utils/pointer-events-freeze.ts [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
 // We apply pointer-events:none on `html` rather than `*` because pointer-events
 // is inherited, so toggling it on a single root element is O(1) style invalidation
 // instead of O(N) for every DOM node, which caused visible lag on dense DOMs
 // like GitHub diff viewers with 10k+ nodes.
+// Same-origin frames forward their input back to the top window, so they can
+// stay interactive for native scrolling without bypassing selection handling.
 // @see https://github.com/aidenybai/react-grab/pull/209
-const POINTER_EVENTS_STYLES = "html { pointer-events: none !important; }";
+const POINTER_EVENTS_STYLES = `html { pointer-events: none !important; }
+iframe[${SAME_ORIGIN_FRAME_ATTRIBUTE}] { pointer-events: auto !important; }`;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/pointer-events-freeze.ts:13]: Fixed in c60a02c1. Accessible frame documents now join the existing pseudo-state freeze lifecycle: each receives its own CSP-aware freeze and hit-test override styles plus hover/focus capture blockers, including frames discovered while already frozen. Frame navigation/removal disposes all styles and listeners. Added a browser regression test proving iframe hover and click side effects stay blocked; the full iframe suite also confirms native iframe scrolling still works while selection is active.

File: packages/react-grab/src/utils/convert-client-position-to-top-window.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    let frameElement: Element | null;
+    try {
+      frameElement = currentWindow.frameElement;
+    } catch {
+      break;
+    }
+    if (!frameElement || !isIframeElement(frameElement)) break;
+
+    const frameBounds = frameElement.getBoundingClientRect();
+    const frameScaleX =
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/convert-client-position-to-top-window.ts]: Fixed in c60a02c1. Frame conversion now uses fractional computed layout dimensions and parsed border and padding insets through getIframeLayoutMetrics instead of integer-rounded offset metrics. The fractional-width iframe regression verifies hit testing and bounds.

File: packages/react-grab/src/utils/convert-parent-position-to-iframe.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  y: number;
+}
+
+export const convertParentPositionToIframe = (
+  iframeElement: HTMLIFrameElement,
+  clientX: number,
+  clientY: number,
+): IframeClientPosition => {
+  const iframeBounds = iframeElement.getBoundingClientRect();
+  const scaleX = iframeElement.offsetWidth > 0 ? iframeBounds.width / iframeElement.offsetWidth : 1;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/convert-parent-position-to-iframe.ts]: This PR intentionally supports axis-aligned positive scaling, which is the transform behavior claimed in the PR summary and covered by the iframe tests. Rotation, skew, and reflection require a different projective-geometry model across the full ancestor chain and are outside this minimal change.

[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/convert-parent-position-to-iframe.ts]: Fixed in c60a02c1 through getIframeLayoutMetrics: scale calculations use fractional computed CSS dimensions and fractional border and padding insets, with offset metrics only as an unavailable-style fallback. The fractional-width iframe regression covers this case.

File: packages/react-grab/src/utils/create-element-bounds.ts [hunk 5]
```diff
@@ -44,13 +46,18 @@ export const createElementBounds = (element: Element): OverlayBounds => {
   }
 
   const rect = element.getBoundingClientRect();
+  const topWindowPosition = convertClientPositionToTopWindow(
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/create-element-bounds.ts:50]: The bounds cache is deliberately limited to one frame (16ms), and relevant frame resize and scroll paths invalidate interaction caches. Animation-only ancestor movement can therefore produce at most one stale frame; retaining that bound is the intended hot-path performance tradeoff.

File: packages/react-grab/src/utils/freeze-pseudo-states.ts [hunk 6]
```diff
@@ -112,12 +115,12 @@ const freezeElement = (
 
 const collectHoveredElements = (cursorX: number, cursorY: number): HTMLElement[] => {
   const hoveredElements: HTMLElement[] = [];
-  let current = document.elementFromPoint(cursorX, cursorY);
+  let current = getDeepElementAtPoint(cursorX, cursorY);
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/freeze-pseudo-states.ts:153]: Fixed in c60a02c1. Frozen inline properties now retain both value and priority, and restoration passes the original priority back to setProperty. The freeze-pseudo-states regression explicitly verifies that !important survives the full freeze and restore cycle.

File: packages/react-grab/src/utils/convert-client-position-to-top-window.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      frameElement.offsetWidth > 0 ? frameBounds.width / frameElement.offsetWidth : 1;
+    const frameScaleY =
+      frameElement.offsetHeight > 0 ? frameBounds.height / frameElement.offsetHeight : 1;
+
+    convertedX = frameBounds.left + (frameElement.clientLeft + convertedX) * frameScaleX;
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/convert-client-position-to-top-window.ts]: This PR intentionally limits frame geometry conversion to axis-aligned positive scaling. Rotated, skewed, or reflected iframe coordinate spaces need full transform-matrix composition and are outside the scale support claimed and tested here.

File: packages/react-grab/src/utils/is-valid-grabbable-element.ts [hunk 8]
```diff
@@ -101,14 +101,16 @@ export const isValidGrabbableElement = (element: Element): boolean => {
   }
 
   const couldBeOverlay =
-    element.clientWidth / window.innerWidth >= VIEWPORT_COVERAGE_THRESHOLD &&
-    element.clientHeight / window.innerHeight >= VIEWPORT_COVERAGE_THRESHOLD;
+    element.clientWidth / (element.ownerDocument.defaultView?.innerWidth ?? window.innerWidth) >=
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/is-valid-grabbable-element.ts:86]: The viewport ratio only gates the expensive overlay style check; it does not reject a viewport-sized element by itself. Ordinary static or relative iframe layout roots remain grabbable. Only fixed or absolute transparent, low-opacity, or high-z-index overlays are filtered, and the iframe viewport is the correct comparison space.

File: packages/react-grab/src/utils/get-deep-elements-at-point.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    const iframeDocument = getAccessibleIframeDocument(topElement);
+    if (iframeDocument) {
+      const iframePosition = convertParentPositionToIframe(topElement, clientX, clientY);
+      collectElementsAtPoint(iframeDocument, iframePosition.x, iframePosition.y, collectedElements);
+    }
+  }
+
+  for (const element of elements) {
+    collectedElements.add(element);
+  }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/get-deep-elements-at-point.ts:36]: Selection fallback is handled by getDeepFallbackElementAtPoint, which walks the z-ordered candidates and descends into each eligible shadow root or accessible iframe. The regression named “descends into an iframe beneath a transparent viewport overlay” covers this exact case. getDeepElementsAtPoint remains the smaller bulk hot path.

File: packages/react-grab/src/utils/observe-same-origin-frame-documents.ts [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+    for (const iframeElement of [...observedIframes.keys()]) {
+      if (iframeElement.ownerDocument === targetDocument) cleanupIframe(iframeElement);
+    }
+
+    shadowRootHookCleanups.get(targetDocument)?.();
+    shadowRootHookCleanups.delete(targetDocument);
+    documentCleanups.get(targetDocument)?.();
+    documentCleanups.delete(targetDocument);
+  };
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/observe-same-origin-frame-documents.ts:41]: Fixed in 3c254dd5. Frame document cleanup now disconnects every observed ShadowRoot owned by the detached document before completing teardown. Added an end-to-end regression that verifies both the frame document and frame shadow observer disconnect.

File: packages/react-grab/src/utils/get-deep-elements-at-point.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    collectElementsAtPoint(topElement.shadowRoot, clientX, clientY, collectedElements);
+  }
+
+  if (topElement && isIframeElement(topElement)) {
+    const iframeDocument = getAccessibleIframeDocument(topElement);
+    if (iframeDocument) {
+      const iframePosition = convertParentPositionToIframe(topElement, clientX, clientY);
+      collectElementsAtPoint(iframeDocument, iframePosition.x, iframePosition.y, collectedElements);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/get-deep-elements-at-point.ts]: Fixed in 3c254dd5. Deep hit-testing now traverses every open ShadowRoot and accessible same-origin iframe in the elementsFromPoint stack, so a transparent overlay no longer hides nested drag targets. Added an end-to-end regression covering both a shadow target and iframe target beneath the overlay.

File: packages/react-grab/src/utils/freeze-animations.ts [hunk 12]
```diff
@@ -165,7 +166,9 @@ const restorePausedAnimations = (animations: Iterable<Animation>): void => {
 const isShadowAnimation = (animation: Animation): boolean => {
   if (!(animation.effect instanceof KeyframeEffect)) return false;
   const target = animation.effect.target;
-  return target instanceof Element && target.getRootNode() instanceof ShadowRoot;
+  if (!isElementNode(target)) return false;
+  const rootNode = target.getRootNode();
+  return isShadowRoot(rootNode) && isReactGrabHost(rootNode.host);
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/freeze-animations.ts:175]: Fixed in e6194c3c. Global animation freeze now tracks each registered same-origin document independently, including frames inserted while already frozen, and cleans up frame-specific CSS, WAAPI, and SVG state on navigation/removal. Added regressions for both pre-existing and late-added iframe animations; focused iframe (29/29), focused coverage (48/48 with 2 retried timing flakes), and the full suite (1,051 passed) are green.

File: packages/react-grab/src/utils/get-elements-in-drag.ts [hunk 13]
```diff
@@ -148,42 +147,33 @@ const filterElementsInDrag = (
     resumePointerEventsFreeze();
   }
 
-  const validCandidates: Element[] = [];
+  const matchingElements: Element[] = [];
   for (const candidateElement of candidates) {
+    if (isIframeElement(candidateElement) && getAccessibleIframeDocument(candidateElement)) {
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/get-elements-in-drag.ts:155]: Fixed in ef32881e. Drag de-duplication now treats an open ShadowRoot boundary specially: the inner candidate replaces its selected host, while ordinary nested candidates are still removed. The regression fixture now gives the host real dimensions, and the focused iframe suite plus the full performance suite pass.

File: packages/react-grab/src/utils/get-deep-elements-at-point.ts [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+          iframeDocument,
+          iframePosition.x,
+          iframePosition.y,
+          collectedElements,
+        );
+      }
+    }
+
+    collectedElements.add(element);
+  }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/get-deep-elements-at-point.ts:36]: Fixed in ef32881e. Deep stack traversal now stops descending into ShadowRoots and same-origin frames once an opaque data-react-grab-ignore layer is encountered, while preserving ordinary lower document layers. Added a drag regression proving the covered iframe target is excluded.

File: packages/react-grab/src/core/context.ts [hunk 15]
```diff
@@ -45,9 +46,9 @@ const isTrustedAppSourcePath = (fileName: string | null | undefined): boolean =>
 const findNearestFiberElement = (element: Element): Element => {
   if (!isInstrumentationActive()) return element;
   let current: Element | null = element;
-  while (current) {
+  while (current?.ownerDocument === element.ownerDocument) {
     if (getFiberFromHostInstance(current)) return current;
-    current = current.parentElement;
+    current = getComposedParentElement(current);
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/core/context.ts]: Fixed in 8855005d. Fiber fallback now follows the physical light-DOM parent chain and only crosses an actual ShadowRoot boundary, so slotted light-DOM nodes cannot latch onto a React-owned slot fiber. Added an E2E fixture with competing slot/light-DOM owners that fails before the fix and resolves to SlottedLightDomOwner after it.

File: packages/react-grab/src/utils/get-deep-fallback-element-at-point.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+      if (iframeDocument) {
+        const iframePosition = convertParentPositionToIframe(candidateElement, clientX, clientY);
+        const iframeTarget = findDeepFallbackElementAtPoint(
+          iframeDocument,
+          iframePosition.x,
+          iframePosition.y,
+        );
+        if (iframeTarget) return iframeTarget;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/react-grab/src/utils/get-deep-fallback-element-at-point.ts:36]: Fixed in 820eaf40. Fallback hit-testing now descends into eligible ShadowRoots and same-origin iframes before applying the grabbability filter to the host candidate, while preserving scope and ignore-layer boundaries. Added an E2E regression with a transparent full-viewport iframe beneath a non-grabbable transparent layer.
